src/java.base/share/classes/java/lang/reflect/ProxyGenerator.java

```diff
@@ -714,100 +723,72 @@ private ProxyMethod(Method method, String sig, Class<?>[] parameterTypes,
          * @param method          The method for which to create a proxy
          * @param methodFieldName the fieldName to generate
          */
-        private ProxyMethod(Method method, String methodFieldName) {
+        private ProxyMethod(Method method, Utf8Entry methodFieldName) {
             this(method, method.toShortSignature(),
-                    method.getSharedParameterTypes(), method.getReturnType(),
-                    method.getSharedExceptionTypes(), method.getDeclaringClass(), methodFieldName);
+                 method.getSharedParameterTypes(), method.getReturnType(),
+                 method.getSharedExceptionTypes(), method.getDeclaringClass(), methodFieldName);
         }
 
         /**
          * Generate this method, including the code and exception table entry.
          */
-        private void generateMethod(ClassWriter cw, String className) {
-            MethodType mt = MethodType.methodType(returnType, parameterTypes);
-            String desc = mt.toMethodDescriptorString();
-            int accessFlags = ACC_PUBLIC | ACC_FINAL;
-            if (method.isVarArgs()) accessFlags |= ACC_VARARGS;
-
-            MethodVisitor mv = cw.visitMethod(accessFlags,
-                    method.getName(), desc, null,
-                    typeNames(Arrays.asList(exceptionTypes)));
-
-            int[] parameterSlot = new int[parameterTypes.length];
-            int nextSlot = 1;
-            for (int i = 0; i < parameterSlot.length; i++) {
-                parameterSlot[i] = nextSlot;
-                nextSlot += getWordsPerType(parameterTypes[i]);
+        private void generateMethod(ClassBuilder clb, ClassEntry className) {
+            var cp = clb.constantPool();
+            var desc = new StringJoiner("", "(", ")" + returnType.descriptorString());
```

Comment:
We should just use an MTD here; the MTD will be passed to StackCounter so we don't have to recompute a MTD.

The MT to MTD conversion shouldn't be too costly; the overhead probably comes from Optional, which we have to wait until Valhalla, as proxy generation is unlikely to be hot and compiled by C2.

Reply:
Original code has been significant in profiler.

```
MethodTypeDesc desc = MethodTypeDesc.of(toClassDesc(returnType),
                    Arrays.stream(parameterTypes).map(ProxyGenerator::toClassDesc).toArray(ClassDesc[]::new));
```

1. each `toClassDesc` builds `descriptorString` and parses/validates it while constructing `ClassDesc`
2. `Arrays.stream(...).map(...).toArray(...)` allocates an array
3. `MethodTypeDesc.of(...)` clones the array and iterates params to check for void
4. `desc.descriptorString()` then finally use the `StringJoiner` 

Optimized code only joins `descriptorString`, no validations, no streaming, no arrays, no cloning.


I suggest this patch as this code is considered as performance critical.
However we can go through `ClassDesc` and `MethodTypeDesc` if not performance critical or if the conversions would be optimized.

For example better (trusted) paths from `MethodType` to `MethodTypeDescriptor` and from `Class` to `ClassDesc`, avoiding at least validations.

Reply:
Turns out your approach to avoid MTD here is apparently useless; `MethodTypeDesc` is still created for initializing the local tracker `topLocal` in `DirectCodeBuilder`. In addition, `StackMapDecoder` also uses `methodTypeSymbol` to compute the initial frame.

IMO we should just stay with MTD; the descriptor breakdown happens too often and, from previous benchmarks, descriptor breakdown is actually slow (which gives CF API a small edge over ASM here). But we can still replace the `parameterList()` iteration with index-based iteration to avoid array copies.

Reply:
Right, there are still so many conversions.
I'll revert the custom code to simplify further optimizations.